cli/internal/globby/globby.go

```diff
@@ -1,55 +1,141 @@
 package globby
 
 import (
-	"os"
+	"errors"
 	"path/filepath"
 	"strings"
 
+	iofs "io/fs"
+
 	"github.com/bmatcuk/doublestar/v4"
 	"github.com/spf13/afero"
 )
 
-var _osFS = afero.NewOsFs()
+var _aferoOsFs = afero.NewOsFs()
+var _aferoIOFS = afero.NewIOFS(_aferoOsFs)
+
+// set is a naive partial implementation of a string set.
+// TODO: remove after we can use generics.
+type set struct {
+	storageMap map[string]bool
+}
+
+// Add a string to the set.
+func (s *set) Add(value string) {
+	s.storageMap[value] = true
+}
+
+// Returns a range of the set's values.
+func (s *set) Values() []string {
+	setValues := make([]string, len(s.storageMap))
+
+	i := 0
+	for key := range s.storageMap {
+		setValues[i] = key
+		i++
+	}
+
+	return setValues
+}
 
 func GlobFiles(basePath string, includePatterns []string, excludePatterns []string) []string {
-	return globFilesFs(_osFS, basePath, includePatterns, excludePatterns)
+	return globFilesFs(_aferoIOFS, basePath, includePatterns, excludePatterns)
 }
 
-func globFilesFs(fs afero.Fs, basePath string, includePatterns []string, excludePatterns []string) []string {
-	var include []string
-	var exclude []string
-	var result []string
+// getRelativePath ensures that the the requested file path is a child of `from`.
+func getRelativePath(from string, to string) (path string, err error) {
+	relativePath, err := filepath.Rel(from, to)
+
+	if err != nil {
+		return "", err
+	}
 
-	for _, p := range includePatterns {
-		include = append(include, filepath.Join(basePath, p))
+	if strings.HasPrefix(relativePath, "..") {
+		return "", errors.New("the path you are attempting to specify is outside of the root")
 	}
 
-	for _, p := range excludePatterns {
-		exclude = append(exclude, filepath.Join(basePath, p))
+	return relativePath, nil
+}
+
+// globFilesFs searches the specified file system to ensure to enumerate all files to include.
+func globFilesFs(fs afero.IOFS, basePath string, includePatterns []string, excludePatterns []string) []string {
+	var processedIncludes []string
+	var processedExcludes []string
+	var result = &set{
+		storageMap: map[string]bool{},
 	}
 
-	includePattern := "{" + strings.Join(include, ",") + "}"
-	excludePattern := "{" + strings.Join(exclude, ",") + "}"
+	for _, includePattern := range includePatterns {
+		includePath := filepath.Join(basePath, includePattern)
+		relativePath, err := getRelativePath(basePath, includePath)
+
+		if err == nil {
```

Comment:
I think we ought to report this error, and in other situations like this one. IIUC, it indicates user error, we were given an unsupported glob.

Reply:
I agree that this represents user error and *should* be reported. I didn't make this change here because I was aiming to keep this PR scoped to "reimplement internals" in order to reduce risk from rewriting a core piece of code. The entirety of this PR's changes exist inside of `globby{,_test}.go` files, and I am aiming to keep it functionally identical to previous behavior.

I'll open a separate PR  as an immediate followup based on this branch later today which adds in error handling and changes the return signature.

Reply:
I have the fast-follow for error plumbing up in #1129.